Added Glances

mcmikemn opened the pull request:
Added Glances, an open-source system cross-platform monitoring tool. It's like a light version of Prometheus.

I added it because Homepage has a widget for it and I wanted to see basic server stats from Homepage. (Homepage also has a widget for Prometheus, but it's far more basic - it's funny: Promethus is a more robsut metrics monitor but the Homepage widget is basic, Glances is a farily basic metrics monitor but the widget is more robust).

EnigmaCurry commented on glances/.env-dist:
```diff
@@ -0,0 +1,51 @@
+GLANCES_MTLS_AUTHORIZED_CERTS=*.clients.glances.example.com
+
+# Timezone
+GLANCES_TZ=EST5EDT
```
Does it support timezones like `America/New_York`? I think that would be a more intuitive example.

EnigmaCurry replied:
link in the comments if appropriate: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

mcmikemn replied:
Fixed in [e86bce8](https://github.com/EnigmaCurry/d.rymcg.tech/pull/284/commits/e86bce8c4914bb96cc0a5d4f9d7427fc8c3c1ac5)

EnigmaCurry commented on glances/docker-compose.instance.yaml:
```diff
@@ -0,0 +1,79 @@
+    labels:
+      #! Services must opt-in to be proxied by Traefik:
+      - "traefik.enable=true"
+
+      #! 'router' is the fully qualified key in traefik for this router/service: project + instance + service
+      #@ router = "{}-{}-{}".format(project,instance,service)
+
+      #! The host matching router rule:
+      - "traefik.http.routers.(@= router @).rule=Host(`(@= traefik_host @)`)"
+      - "traefik.http.routers.(@= router @).entrypoints=websecure"
+
+      #@ enabled_middlewares.append("{}-ipallowlist".format(router))
+      - "traefik.http.middlewares.(@= router @)-ipallowlist.ipallowlist.sourcerange=(@= ip_sourcerange @)"
+      #@ if enable_http_auth:
+      #@ enabled_middlewares.append("{}-basicauth".format(router))
+      - "traefik.http.middlewares.(@= router @)-basicauth.basicauth.users=(@= http_auth @)"
+      - "traefik.http.middlewares.(@= router @)-basicauth.basicauth.headerField=X-Forwarded-User"
+      #@ end
+
+      #@ if enable_oauth2:
+      #@ enabled_middlewares.append("traefik-forward-auth@docker")
+      #@ enabled_middlewares.append("header-authorization-group-{}@file".format(authorized_group))
+      #@ end
+
+      #@ if enable_mtls_auth:
+      - "traefik.http.routers.(@= router @).tls.options=step_ca_mTLS@file"
+        #@ if len(mtls_authorized_certs):
+      - "traefik.http.middlewares.mtlsauth-(@= router @).plugin.certauthz.domains=(@= mtls_authorized_certs @)"
+        #@ enabled_middlewares.append("mtlsauth-{}".format(router))
+        #@ end
+        #@ enabled_middlewares.append("mtls-header@file")
+      #@ end
+      
+      #! Override the default port that the app binds to:
+      #! You don't normally need to do this, as long as your image has
+      #! an EXPOSE directive in it, Traefik will autodetect it, but this is how you can override it:
+      - "traefik.http.services.(@= router @).loadbalancer.server.port=61208"
+
+      #! Apply all middlewares (do this at the end!)
+      - "traefik.http.routers.(@= router @).middlewares=(@= ','.join(enabled_middlewares) @)"
```
Glances exposes two services, should there be a second router for the XMLRPC service? 

https://github.com/nicolargo/glances/blob/24c87253595f9d3168afc26e82cfa464fc7523b9/docker-files/ubuntu.Dockerfile#L101

mcmikemn replied:
I don't know. It works for me with just port 61208. Maybe XMLRPC is used for other funtions I'm not using (e.g., Glances has the ability to report on multiple servers - maybe it uses XMLRPC to communicate between Glances instances?)?

`docker ps` displays this container with no ports:
```
<containerId>   nicolargo/glances:4.1.2.1                                "/bin/sh -c '/venv/b…"   2 hours ago    Up 2 hours                                  glances-glances-1
```

EnigmaCurry replied:
This is the behaviour of `network_mode: host` same as traefik. In effect, Docker has no clue what your container is doing network wise, because its taken over the host network.

EnigmaCurry replied:
so this means that probably 61209 is probably already open, but your external firewall probably blocks it.

EnigmaCurry commented on glances/docker-compose.yaml:
```diff
@@ -0,0 +1,15 @@
+    image: ${GLANCES_IMAGE}
+    restart: unless-stopped
+    pid: host
+    privileged: true
```
We should not use privileged: true, instead follow SECURITY.md and find the minimum capabilities required.

EnigmaCurry replied:
https://github.com/EnigmaCurry/d.rymcg.tech/blob/master/SECURITY.md#drop-system-capabilities

mcmikemn replied:
Fixed in [e86bce8](https://github.com/EnigmaCurry/d.rymcg.tech/pull/284/commits/e86bce8c4914bb96cc0a5d4f9d7427fc8c3c1ac5)

EnigmaCurry commented on glances/docker-compose.yaml:
```diff
@@ -0,0 +1,15 @@
+    privileged: true
+    network_mode: "host"
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock:ro
```
Is docker required for all functionality?

mcmikemn replied:
I think Glances only uses docker socket access (read-only) to report metrics of containers. My personal use-case is for servers, not containers, so I could remove docker socket access. Or we could variablize it and leave it up to the user?

EnigmaCurry replied:
Unfortunately the `ro` is completely and total LIE. This gives read and write privileges.

EnigmaCurry replied:
In the case of homepage we made it optional, because not everything in homepage requires it. So I would prefer that if there are still usable features of glances that work without it.

mcmikemn replied:
If we decide to keep Glances in d.ry, should we just eliminate the docker socker access?

mcmikemn replied:
Fixed in [e86bce8](https://github.com/EnigmaCurry/d.rymcg.tech/pull/284/commits/e86bce8c4914bb96cc0a5d4f9d7427fc8c3c1ac5)

EnigmaCurry commented on glances/docker-compose.yaml:
```diff
@@ -0,0 +1,15 @@
+    restart: unless-stopped
+    pid: host
+    privileged: true
+    network_mode: "host"
```
Why is network mode host? This is insecure for traefik.

mcmikemn replied:
ChatGPT tells me Glances needs to monitor the host's network interfaces. Using host network mode enables it to see and track real-time traffic on the host's network interfaces without being isolated within a container-specific virtual network. And if you are running other services on the host machine that Glances needs to monitor, such as local databases or web servers, host mode allows the container to interact with them directly without needing complex network configurations or port mappings.

I installed without `network_mode: "host"` and Homepage can't see Glances (or more likely, Glances can't see the host so it has nothing to report).

EnigmaCurry replied:
ok maybe I need to relent on both of these issues, and just let it do its thing. We need a huge warning on the top of the README that this is essentially a system root level container.

EnigmaCurry replied:
```
## Warning

This container uses the docker flags `privileged: true` and
`network_mode: host`. This gives Glances unlimited access to your host
system. You should not install this unless you completely trust this
service.
```

EnigmaCurry commented:
https://github.com/nicolargo/glances/tree/24c87253595f9d3168afc26e82cfa464fc7523b9?tab=readme-ov-file#docker-the-cloudy-way

This example does use network mode host, but not privileged. Is privileged necessary?

EnigmaCurry commented:
```
## Warning

This container uses the docker flag `network_mode: host`. This gives Glances unlimited access to your host
network. You should not install this unless you completely trust this service. To enforce the use of traefik, your external firewall should block TCP ports 61208 and 61209.
```

EnigmaCurry commented:
With the addition of the README warning above, I am happy with this as-is.

mcmikemn commented:
I reduced priveleges, added alert, variablized bind mount to docker socket, updated readme.

You were already OK with this PR with only the alert added. Do you want to review my changes, @EnigmaCurry?

EnigmaCurry commented on glances/docker-compose.instance.yaml:
```diff
@@ -37,6 +38,13 @@ services:
+      - /var/run/docker.sock:/var/run/docker.sock:ro
+      #@ end
+      #! The following line causes Glances to display host's OS detail instead of container's
+      - /etc/os-release:/etc/os-release:ro
```
Will you also test if you can do the timezone and localtime as shown in #285 and then you probably dont need the TZ env var at all.

mcmikemn replied:
Didn't work. We decided to approve this PR using $TZ.
